net/dns/resolved.go

```diff
@@ -74,73 +76,164 @@ type resolvedLinkDomain struct {
+	res    chan<- error // response channel
+}
+
 // resolvedManager is an OSConfigurator which uses the systemd-resolved DBus API.
 type resolvedManager struct {
+	ctx    context.Context
+	cancel func() // terminate the context, for close
+
 	logf  logger.Logf
 	ifidx int
 
-	cancelSyncer context.CancelFunc // run to shut down syncer goroutine
-	syncerDone   chan struct{}      // closed when syncer is stopped
-	resolved     dbus.BusObject
-
-	mu     sync.Mutex // guards RPCs made by syncLocked, and the following
-	config OSConfig   // last SetDNS config
+	configCR chan changeRequest // tracks OSConfigs changes and error responses
 }
 
 func newResolvedManager(logf logger.Logf, interfaceName string) (*resolvedManager, error) {
-	conn, err := dbus.SystemBus()
-	if err != nil {
-		return nil, err
-	}
-
 	iface, err := net.InterfaceByName(interfaceName)
 	if err != nil {
 		return nil, err
 	}
 
 	ctx, cancel := context.WithCancel(context.Background())
 
-	ret := &resolvedManager{
-		logf:         logf,
-		ifidx:        iface.Index,
-		cancelSyncer: cancel,
-		syncerDone:   make(chan struct{}),
-		resolved:     conn.Object(dbusResolvedObject, dbus.ObjectPath(dbusResolvedPath)),
-	}
-	signals := make(chan *dbus.Signal, 16)
-	go ret.resync(ctx, signals)
-	// Only receive the DBus signals we need to resync our config on
-	// resolved restart. Failure to set filters isn't a fatal error,
-	// we'll just receive all broadcast signals and have to ignore
-	// them on our end.
-	if err := conn.AddMatchSignal(dbus.WithMatchObjectPath(dbusPath), dbus.WithMatchInterface(dbusInterface), dbus.WithMatchMember(dbusOwnerSignal), dbus.WithMatchArg(0, dbusResolvedObject)); err != nil {
-		logf("[v1] Setting DBus signal filter failed: %v", err)
+	mgr := &resolvedManager{
+		ctx:    ctx,
+		cancel: cancel,
+
+		logf:  logf,
+		ifidx: iface.Index,
+
+		configCR: make(chan changeRequest),
 	}
-	conn.Signal(signals)
-	return ret, nil
+
+	go mgr.run(ctx)
+
+	return mgr, nil
 }
 
 func (m *resolvedManager) SetDNS(config OSConfig) error {
-	m.mu.Lock()
-	defer m.mu.Unlock()
+	errc := make(chan error, 1)
+	defer close(errc)
+
+	select {
+	case <-m.ctx.Done():
+		return m.ctx.Err()
+	case m.configCR <- changeRequest{config, errc}:
 
-	m.config = config
-	return m.syncLocked(context.TODO()) // would be nice to plumb context through from SetDNS
+	}
+
+	select {
+	case <-m.ctx.Done():
+		return m.ctx.Err()
+	case err := <-errc:
+		m.logf("Resolved systemBus connection interrupted with error: %v.", err)
+		return err
+	}
 }
 
-func (m *resolvedManager) resync(ctx context.Context, signals chan *dbus.Signal) {
-	defer close(m.syncerDone)
+func (m *resolvedManager) run(ctx context.Context) {
+	var (
+		conn      *dbus.Conn
+		signals   chan *dbus.Signal
+		r1Manager dbus.BusObject
+	)
+	bo := backoff.NewBackoff("resolvedSetDNS", m.logf, 30*time.Second)
+	needsReconnect := make(chan bool, 1)
+	defer close(needsReconnect)
+	defer close(m.configCR)
```

Comment:
same. But this one's arguably worse because this goroutine is the reader of this channel and close is a write operation. So you can be closing a channel concurrently with a SetDNS call which would then panic

best to just remove this one